rust-runtime/inlineable/src/endpoint_lib/partition.rs

```diff
@@ -0,0 +1,577 @@
+//! Partition function to determine a partition for a given region
+//!
+//! This function supports adding regions dynamically, parsing a JSON file, and builder construction.
+//!
+//! If, at a future point, this interface stabilizes it is a good candidate for extraction into a
+//! shared crate.
+use crate::endpoint_lib::diagnostic::DiagnosticCollector;
+use aws_smithy_json::deserialize::Error;
+use regex::Regex;
+use std::borrow::Cow;
+use std::collections::HashMap;
+
+/// Determine the AWS partition metadata for a given region
+pub(crate) struct PartitionResolver {
+    partitions: Vec<PartitionMetadata>,
+}
+
+impl PartitionResolver {
+    pub(crate) fn from_partitions(partitions: Vec<PartitionMetadata>) -> Self {
+        Self { partitions }
+    }
+}
+
+/// Partition result returned from partition resolver
+pub(crate) struct Partition<'a> {
+    name: &'a str,
+    dns_suffix: &'a str,
+    dualstack_dns_suffix: &'a str,
+    supports_fips: bool,
+    supports_dualstack: bool,
+}
+
+static DEFAULT_OVERRIDE: &PartitionOutputOverride = &PartitionOutputOverride {
+    name: None,
+    dns_suffix: None,
+    dualstack_dns_suffix: None,
+    supports_fips: None,
+    supports_dualstack: None,
+};
+
+/// Merge the base output and the override output, dealing with Cow's
+macro_rules! merge {
+    ($base: expr, $output: expr, $field: ident) => {
+        $output
+            .$field
+            .as_ref()
+            .map(|s| s.as_ref())
+            .unwrap_or($base.outputs.$field.as_ref())
+    };
+}
+
+impl PartitionResolver {
+    pub(crate) fn new() -> PartitionResolver {
+        PartitionResolver { partitions: vec![] }
+    }
+    pub(crate) fn add_partition(&mut self, partition: PartitionMetadata) {
+        self.partitions.push(partition);
+    }
+
+    /// Resolve a partition for a given region
+    ///
+    /// 1. Enumerate each partition in the `partitions` array, and determine if the identifier to be resolved matches an
+    ///    explicit
+    ///    region listed in the `regions` array for a given partition. If identifier matches, proceed to step 4, otherwise
+    ///    continue to step 2.
```

Comment:
```suggestion
    /// 1. Enumerate each partition in the `partitions` array, and determine if the identifier to be resolved matches an
    ///    explicit region listed in the `regions` array for a given partition. If identifier matches, proceed to step 4, 
    ///    otherwise continue to step 2.
```